members/views/user/signup.py

```diff
@@ -13,9 +13,9 @@
 @transaction.atomic
 def signup(request, member_id):
     member = get_object_or_404(Member, pk=member_id)
-    if member.invitation_code_end_date is None or member.invitation_code_end_date < timezone.now():
-        return HttpResponse("Code is no longer valid")
     if not request.user.has_perm('auth.add_user'):
+        if member.invitation_code_end_date is None or member.invitation_code_end_date < timezone.now():
+            return HttpResponse("Code is no longer valid")
         members = Member.objects.filter(pk=member_id, invitation_code=request.GET.get('key', ''))
 
         if len(members) != 1:
```

Comment:
Two lines below this one the check is still wrong I think?

Reply:
I don't think so the check that you get only one member from the member_id + invitation_code seems legit. I would expect you get only one. And it is also a check that there is at least one member with this invitation code.

Reply:
There does not have to be an invitation code if you create an directly (using the `new user` button)

Reply:
Generally, I'd rewrite this function into something like this:
1. Get member for ID
2. If code is not empty: do checks based on the code
3. If code is empty, give error if member doesn't have permission for editing stuff without the link.

This way the logic is less likely to get mistakes later.

Reply:
I prefer to first check if the user can edit without the link because that logic path is then done. This makes the code a bit simpler as the other checks don't get polluted by the permission check

Reply:
The current logic is strange and very unexpected though.
If you use an expired link while you are logged in (and you're an admin), then the link suddenly works, which doesn't really make sense to me.
 It also has strange negations, and a totally needless lookup from the database as it currently is.
Compare it to
```
member =  get_object_or_404(Member, pk=member_id)
invitation_code=request.GET.get('key', '')
if invitation_code == "":
  if not request.user.has_perm('auth.add_user'):
    raise WhateverException
else:
  if member.invitation_code != invitation_code:
    return HttpResponse("Wrong invite code")
  if not member.invitation_code_valid:
        return HttpResponse("Member has no valid invitation")
   if member.invitation_code_end_date is None or member.invitation_code_end_date < timezone.now():
      return HttpResponse("Code is no longer valid")
# Rest of code
```

Reply:
Ahh I get it now. I did not think of the unexpected behaviour part makes sense now. I will change it